src/utils/str.c

```diff
@@ -169,31 +257,33 @@ char str_at(const str *string, size_t pos) {
     return string->data[pos];
 }
 
-void str_toupper(str *string) {
-    for(size_t i = 0; i < str_size(string); i++) {
-        string->data[i] = toupper(string->data[i]);
-    }
-}
-
-void str_tolower(str *string) {
-    for(size_t i = 0; i < str_size(string); i++) {
-        string->data[i] = tolower(string->data[i]);
-    }
-}
+// ------------------------ Type conversion ------------------------
 
-int str_to_int(const str *string, int *result) {
-    *result = atoi(str_c(string));
-    return 1;
+bool str_to_float(const str *string, float *result) {
+    char *end;
+    *result = strtof(string->data, &end);
+    return (string->data != end);
 }
 
-int str_to_float(const str *string, float *result) {
-    *result = atof(str_c(string));
-    return 1;
+bool str_to_long(const str *string, long *result) {
+    char *end;
+    *result = strtol(string->data, &end, 10);
+    return (string->data != end);
 }
 
-int str_to_long(const str *string, long *result) {
-    *result = atol(str_c(string));
-    return 1;
+bool str_to_int(const str *string, int *result) {
+    char *end;
+    long out = strtol(string->data, &end, 10);
+    if (string->data == end) {
+        *result = 0;
+        return false;
+    }
+    if (out > INT_MAX || out < INT_MIN) {
+        *result = 0;
+        return false;
+    }
+    *result = out;
```

Comment:
i guess explicit cast would be ok to add here

Reply:
Actually, I think I will just drop the str_to_int function, since it doesn't seem to be used at all anymore. I'll just leave the str_to_float and str_to_long since they are simple.

Reply:
Dropped.